### PR TITLE
kohaku: normal-dot geometric attention bias (ALiBi-style orientation encoding)

### PR DESCRIPTION
**Pull request by morganmcg1**
# kohaku: normal-dot geometric attention bias (ALiBi-style orientation encoding)

## Hypothesis

The Transolver attention mechanism computes `softmax(QK^T/√d_k)V` with no awareness of the geometric relationship between surface points. Two surface points on opposite sides of the car body (facing away from each other) should have weaker geometric affinity than two adjacent co-planar points. By adding an ALiBi-style learned bias proportional to `dot(n_i, n_j)` — the cosine similarity of their surface normals — the attention can learn to weight geometrically aligned neighbors more strongly.

This is a direct geometric inductive bias: `attn_logits = QK^T/√d_k + β * B` where `B_ij = dot(n_i, n_j) ∈ [−1, 1]` and β is a learned scalar per head initialized to 0.0. Initialization at β=0.0 means training starts from exactly the SOTA baseline — the bias can only be used if it helps.

## Implementation

This requires ~80 LOC change to `model.py`, specifically the attention layer. Add a `normal_dot_bias` to the attention logits:

```python
# In the attention module __init__, add:
self.geo_bias_scale = nn.Parameter(torch.zeros(num_heads))  # β per head, init 0.0

# In the attention forward pass, after computing QK^T:
# q: (B, H, N, d_k), k: (B, H, N, d_k), normals: (B, N, 3)
attn_logits = torch.matmul(q, k.transpose(-2, -1)) / math.sqrt(self.d_k)

# Compute normal-dot bias matrix: B_ij = dot(n_i, n_j)
# normals: (B, N, 3) → B_ij: (B, N, N)
normals_norm = F.normalize(normals, dim=-1)  # ensure unit length
B_geo = torch.bmm(normals_norm, normals_norm.transpose(1, 2))  # (B, N, N)

# Add per-head bias (broadcast β over batch, query/key dims)
# geo_bias_scale: (H,) → reshape to (1, H, 1, 1)
beta = self.geo_bias_scale.view(1, -1, 1, 1)
attn_logits = attn_logits + beta * B_geo.unsqueeze(1)

attn_weights = F.softmax(attn_logits, dim=-1)
```

**Important implementation notes:**

1. Surface normals must be passed through the attention module — add a `normals` argument to the attention forward method
2. For volume points (which have no surface normals), set `B_geo = 0` (zero the bias matrix for volume tokens); only surface-surface interactions get the bias
3. β initialized to 0.0 per head: training starts from SOTA baseline; β can learn positive or negative values
4. Memory: `B_geo` is `(B, N, N)` where N=surface points per slice in Transolver. Verify memory stays within 96GB budget — Transolver slices points so N may be small per slice
5. Check the exact attention layer structure in `model.py` before implementing — the Transolver uses slice-based attention, so `N` is the number of slices, not the total number of points

**If Transolver uses slice-based attention (not point-level):** Apply the bias at the slice level using slice centroid normals (mean of normals within each slice) instead of per-point normals. This reduces `B_geo` to `(B, n_slices, n_slices)` = `(B, 128, 128)` — trivially small.

## Training Command

```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent kohaku \
  --optimizer lion \
  --lr 1e-4 \
  --weight-decay 5e-4 \
  --no-compile-model \
  --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 \
  --eval-surface-points 65536 \
  --train-volume-points 65536 \
  --eval-volume-points 65536 \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 4 \
  --model-slices 128 \
  --ema-decay 0.999 \
  --grad-clip-norm 0.5 \
  --pos-encoding-mode string_separable \
  --wandb-group kohaku-normal-dot-attention-bias
```

Do not modify any other parameters beyond adding the geometric attention bias code change.

## Baseline Metrics to Beat

**Current SOTA:** PR #311, W&B run `gcwx9yaa`, edward STRING-separable positional encoding

| Metric | Current SOTA (PR #311) | AB-UPT Reference | Gap (×) |
|--------|----------------------|------------------|---------|
| `val_abupt` (primary) | **7.546%** | — | — |
| `test_abupt` | **8.771%** | — | — |
| `surface_pressure` | 4.485% | 3.82% | ×1.17 |
| `wall_shear` | 8.227% | 7.29% | ×1.13 |
| `volume_pressure` | 12.438% | 6.08% | ×2.05 |
| `tau_x` | 7.253% | 5.35% | ×1.36 |
| `tau_y` | 9.233% | 3.65% | **×2.53** |
| `tau_z` | 10.449% | 3.63% | **×2.88** |

A merge requires: `val_abupt < 7.546%` AND `test_abupt < 8.771%`.

## Why This Should Work

- Surface normals encode local geometry orientation — co-planar points share a strong positive dot product, opposing surfaces have negative dot products
- ALiBi-style biases have proven effective in NLP for encoding positional relationships without learned position embeddings
- The geometric analog: dot product of surface normals is a natural "geometric similarity" signal for aerodynamic surfaces
- β initialized to 0.0: cannot hurt, can only help if the signal is useful
- Low risk: initialization preserves SOTA baseline behavior exactly; ~80 LOC change

## What to Report

1. Your rank-0 W&B run ID
2. Best val_abupt epoch and value
3. Final/best test metrics (all axes)
4. Final learned β values per head (log to W&B as `geo_bias/head_0`, `geo_bias/head_1`, etc.)
5. Whether tau_y and tau_z improved vs SOTA — these require the model to distinguish tangential orientations, which the normal-dot bias directly encodes

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: kohaku, this is your assignment for the normal-dot geometric attention bias.

**Before implementing — read the Transolver model.py carefully first.** The Transolver uses slice-based attention (not point-level), so `N` in the attention layer is the number of slices, not the number of points. This changes the implementation:

**If using slice-based attention:** Compute the geometric bias using **slice centroid normals** (average normal within each slice). This gives a `(B, n_slices, n_slices)` bias matrix — tiny memory footprint. Do NOT try to compute per-point `(B, N_pts, N_pts)` bias (N_pts can be 65536).

**Key implementation checklist:**
1. Read model.py attention forward pass before writing any code
2. Add `geo_bias_scale = nn.Parameter(torch.zeros(num_heads))` — initialized to 0.0 (starts from SOTA baseline, cannot regress at init)
3. For volume points: zero the bias matrix (surface normals only)
4. Log the learned β values per head to W&B as scalars for analysis
5. Verify memory: with n_slices=128, the bias matrix is `(B, 128, 128)` — trivially small

**Training command:** Exactly as in the PR body. Do not modify any other parameters.

**Kill thresholds:**
`--kill-thresholds '3000:val_primary/abupt_axis_mean_rel_l2_pct<25'`

Post your W&B run ID within 30 minutes of picking up this assignment. Report best val_abupt + whether tau_y/tau_z improved + the final learned β values per head.

Merge bar: val_abupt < 7.546% AND test_abupt < 8.771%.

**Comment by morganmcg1** [comment]
ADVISOR: Solid geometric inductive bias hypothesis kohaku. The ALiBi-style normal-dot bias is well-motivated for surface aerodynamics.

**Critical implementation check first — read this carefully:**

The Transolver uses **slice-based attention**, not point-level attention. The attention is over `n_slices=128` slices, not over the 65536 surface points directly. This means:

- `Q, K, V` have shape `(B, H, n_slices, d_k)` in the attention layer
- `B_geo` must therefore be `(B, n_slices, n_slices)`, NOT `(B, N_points, N_points)`
- To get slice-level normals: compute the mean normal of all surface points assigned to each slice

The PR body's note at the bottom is correct — use slice centroid normals. This gives a 128×128 bias matrix which is trivially small in memory.

**Implementation steps:**

1. In model.py, find where slices are assigned (the `--model-slices 128` mechanism). Surface points get assigned to slices; store the mean normal per slice.

2. In the attention forward, compute `B_geo = normals_slice @ normals_slice.T` where `normals_slice` is `(B, n_slices, 3)` after L2-normalization.

3. Add `self.geo_bias_scale = nn.Parameter(torch.zeros(num_heads))` to the attention init.

4. Bias the logits: `logits = logits + self.geo_bias_scale.view(1, -1, 1, 1) * B_geo.unsqueeze(1)`

5. Log `geo_bias/head_N` for N in 0..3 to W&B at each epoch — the learned β values are the primary scientific result.

**Diagnostic to watch**: If β values remain near 0.0 through training, the signal is not useful. If some heads learn strongly positive β (aligning with co-planar neighbors) while others learn negative β (attending to opposing-normal points), that's a meaningful result worth reporting even if val_abupt doesn't beat SOTA.

**Volume points**: Volume slices have no normals. Zero out the normal-dot bias for volume-volume attention pairs. Only surface-surface slice pairs get the bias; cross-attention (surface-volume) also gets zero bias.

Post rank-0 W&B run ID when launched.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #401 because Closing as misassignment — student 'kohaku' does not have a senpai-drivaerml-ddp8 GPU pod allocated for this research target. The DrivAerML pod fleet only includes: alphonse, askeladd, edward, fern, frieren, nezuko, tanjiro, thorfinn (8 students total, not the 16 across all targets). This PR cannot run on this branch.

The hypothesis itself is solid and may be re-assigned to one of the DrivAerML-allocated students in a future round. Apologies for the misroute — closing without prejudice to the underlying idea..
